core/drivers/regulator/regulator_dt.c

```diff
@@ -0,0 +1,368 @@
+
+	cuint = fdt_getprop(fdt, node, "regulator-name", NULL);
+	if (cuint) {
+		/* Replace name with the one found from the DT node */
+		char *name = (char *)cuint;
+
+		free(regulator->name);
+		regulator->name = strdup(name);
+		if (!regulator->name)
+			return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	cuint = fdt_getprop(fdt, node, "regulator-min-microvolt", &len);
+	if (cuint && len == sizeof(*cuint))
+		regulator->min_uv = fdt32_to_cpu(*cuint);
+	else if (cuint || len != -FDT_ERR_NOTFOUND)
+		panic();
+
+	cuint = fdt_getprop(fdt, node, "regulator-max-microvolt", &len);
+	if (cuint && len == sizeof(*cuint)) {
+		regulator->max_uv = fdt32_to_cpu(*cuint);
+
+		if (regulator->max_uv < regulator->min_uv) {
+			EMSG("Regulator %s max_uv %d < %d",
+			     regulator_name(regulator), regulator->max_uv,
+			     regulator->min_uv);
+
+			return TEE_ERROR_GENERIC;
+		}
+	} else if (cuint || len != -FDT_ERR_NOTFOUND) {
+		panic();
+	} else {
+		regulator->max_uv = INT_MAX;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * API function to register a DRIVER_REGULATOR provider instance.
+ * The registration can be deferred if the regulator supply (if any)
+ * is not yet registered, in which case the regulator is placed in
+ * a regulator pending list.
+ */
+TEE_Result regulator_dt_register(const void *fdt, int node, int provider_node,
+				 const struct regu_dt_desc *desc)
+{
+	TEE_Result res = TEE_ERROR_OUT_OF_MEMORY;
+	struct regulator *regulator = NULL;
+	uint32_t supply_phandle = 0;
+	char *name = NULL;
+
+	assert(desc);
+
+	if (!desc->regulator) {
+		regulator = calloc(1, sizeof(*regulator));
+		if (!regulator)
+			return TEE_ERROR_OUT_OF_MEMORY;
+	} else {
+		regulator = desc->regulator;
+		memset(regulator, 0, sizeof(*regulator));
+	}
+
+	if (desc->name) {
+		/* Will be freed if overridden by DT node content */
+		name = strdup(desc->name);
+		if (!name)
+			goto err_free;
+	}
+
+	*regulator = (struct regulator){
+		.name = name,
+		.ops = desc->ops,
+		.priv = desc->priv,
+	};
+
+	res = parse_dt(fdt, node, regulator);
+	if (res)
+		goto err_free;
+
+	if (desc->supply_name) {
+		res = get_supply_phandle(fdt, provider_node, desc->supply_name,
+					 &supply_phandle);
+		if (res)
+			goto err_free;
+	}
+
+	if (supply_phandle) {
+		res = add_to_pending_list(fdt, node, regulator, supply_phandle,
+					  !desc->regulator);
+		if (res)
+			goto err_free;
+	} else {
+		res = register_final(fdt, node, regulator);
+		if (res)
+			goto err_free;
+	}
+
+	resolve_pending_list();
+
+	return TEE_SUCCESS;
+
+err_free:
+	/* This function cannot return TEE_ERROR_DEFER_DRIVER_INIT */
+	assert(res != TEE_ERROR_DEFER_DRIVER_INIT);
+
+	free(regulator->name);
+	if (!desc->regulator)
+		free(regulator);
+
+	return res;
+}
+
+static TEE_Result release_regulator_pending_lists(void)
+{
+	struct pending_regu *pending = NULL;
+	struct pending_regu *next = NULL;
+
+	if (!SLIST_EMPTY(&pending_regu_list))
+		DMSG("Some regualtors were not supplied:");
```

Comment:
regulators

Reply:
thanks. it's strange that checkpatch did not warn!